src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -213,129 +150,58 @@ private void SendAsync(IPAddress address, int timeout, byte[] buffer, PingOption
             CheckStart();
             try
             {
-                _asyncOp = AsyncOperationManager.CreateOperation(userToken);
-                InternalSendAsync(addressSnapshot, buffer, timeout, options);
+                return SendPingAsyncCore(addressSnapshot, buffer, timeout, options);
             }
             catch (Exception e)
             {
                 Finish();
-                throw new PingException(SR.net_ping, e);
+                return Task.FromException<PingReply>(new PingException(SR.net_ping, e));
             }
         }
 
-        public Task<PingReply> SendPingAsync(IPAddress address)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(address, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(string hostNameOrAddress)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(hostNameOrAddress, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(IPAddress address, int timeout)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(address, timeout, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(string hostNameOrAddress, int timeout)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(hostNameOrAddress, timeout, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(IPAddress address, int timeout, byte[] buffer)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(address, timeout, buffer, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(string hostNameOrAddress, int timeout, byte[] buffer)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(hostNameOrAddress, timeout, buffer, tcs));
-        }
-
-        public Task<PingReply> SendPingAsync(IPAddress address, int timeout, byte[] buffer, PingOptions options)
-        {
-            return SendPingAsyncCore(tcs => SendAsync(address, timeout, buffer, options, tcs));
-        }
-
         public Task<PingReply> SendPingAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions options)
         {
-            return SendPingAsyncCore(tcs => SendAsync(hostNameOrAddress, timeout, buffer, options, tcs));
-        }
-
-        private Task<PingReply> SendPingAsyncCore(Action<TaskCompletionSource<PingReply>> sendAsync)
-        {
-            // Create a TaskCompletionSource to represent the operation.
-            var tcs = new TaskCompletionSource<PingReply>();
-
-            // Register a handler that will transfer completion results to the TCS Task.
-            PingCompletedEventHandler handler = null;
-            handler = (sender, e) => HandleCompletion(tcs, e, handler);
-            PingCompleted += handler;
+            if (string.IsNullOrEmpty(hostNameOrAddress))
+            {
+                throw new ArgumentNullException("hostNameOrAddress");
```

Comment:
Is this exception a real user scenario or an internal `Assert` kind of thing?  If this exception gets to the dev caller, then will then be able to map the `hostNameOrAddress` to some public API that they called? It might be better to do public API parameter validation in the public API layer itself.